shared/js/content-scripts/block-cookie.js

```diff
@@ -0,0 +1,143 @@
+                if (event && event.isTrusted && event.data && event.data.source === secret) {
+                    if (event.data.type === 'tracker') {
+                        trackerHosts.add(event.data.hostname)
+                    } else {
+                        resolve(event.data)
+                    }
+                }
+            }
+            window.addEventListener('message', messageListener)
+        })
+        document.__defineSetter__('cookie', (value) => {
+            // call the native document.cookie implementation. This will set the cookie immediately
+            // if the value is valid. We will override this set later if the policy dictates that
+            // the expiry should be changed.
+            cookieSetter.apply(document, [value])
+            try {
+                // determine the origins of the scripts in the stack
+                const stack = new Error().stack.split('\n')
+                const scriptOrigins = stack.reduce((origins, line) => {
+                    const res = line.match(lineTest)
+                    if (res && res[2]) {
+                        origins.add(new URL(res[2]).hostname)
+                    }
+                    return origins
+                }, new Set())
+
+                // wait for config before doing same-site tests
+                loadPolicy.then(({ shouldBlock, tabRegisteredDomain, policy, isTrackerFrame }) => {
+                    if (!tabRegisteredDomain || !shouldBlock) {
+                        // no site domain for this site to test against, abort
+                        debug && console.log('[ddg-cookie-policy] policy disabled on this page')
+                        return
+                    }
+                    const sameSiteScript = [...scriptOrigins].every((host) => host === tabRegisteredDomain || host.endsWith(`.${tabRegisteredDomain}`))
+                    if (sameSiteScript) {
+                        // cookies set by scripts loaded on the same site as the site are not modified
+                        debug && console.log('[ddg-cookie-policy] ignored (sameSite)', value, [...scriptOrigins])
+                        return
+                    }
+                    const trackerScript = [...scriptOrigins].some((host) => trackerHosts.has(host))
+                    if (!trackerScript && !isTrackerFrame) {
+                        debug && console.log('[ddg-cookie-policy] ignored (non-tracker)', value, [...scriptOrigins])
+                        return
+                    }
+                    // extract cookie expiry from cookie string
+                    const cookie = new Cookie(value)
+                    // apply cookie policy
+                    if (cookie.getExpiry() > policy.threshold) {
+                        cookie.maxAge = policy.maxAge
+                        debug && console.log('[ddg-cookie-policy] update', cookie.toString(), scriptOrigins)
+                        cookieSetter.apply(document, [cookie.toString()])
```

Comment:
there is a slight chance for a race condition where cookie was removed in the meantime and we are recreating it. Maybe it's worth checking if it's still in `document.cookie` before we write?

Reply:
Good catch. I'll add a check.

Reply:
Fixed by https://github.com/duckduckgo/duckduckgo-privacy-extension/pull/570/commits/862fbf6a595bc66f17d272b9c17ab4c2288b6227